File: packages/python/wood_opt/optimizer1d.py
from __future__ import annotations
from typing import List, Optional
from .types import (
    Stock1D,
    Part1D,
    MinRemnantSize,
    OptimizationResult,
    StockResult1D,
    Placement1D,
    Cut1D,
    Segment1D,
    Summary,
    UnplacedPart,
)
from .binpacking import (
    BinDefinition,
    ItemDefinition,
    BinPacking1DOptions,
    bin_pack_1d,
)


def optimize_1d(
    stocks: List[Stock1D],
    parts: List[Part1D],
    kerf: float = 0.0,
    min_remnant_size: Optional[MinRemnantSize] = None,
) -> OptimizationResult:
    kerf = max(0.0, float(kerf))
    min_remnant_length = float(min_remnant_size.length) if (min_remnant_size and min_remnant_size.length is not None) else 0.0

    bins = [
        BinDefinition(
            id=s.id,
            capacity=float(s.length),
            quantity=s.quantity if s.quantity is not None else 1,
            cost=float(s.cost) if s.cost is not None else float(s.length),
            data=s,
        )
        for s in stocks
    ]

    items = [
        ItemDefinition(
            id=p.id,
            size=float(p.length),
            quantity=p.quantity if p.quantity is not None else 1,
            data=p,
        )
        for p in parts
    ]

    pack_result = bin_pack_1d(
        bins=bins,
        items=items,
        options=BinPacking1DOptions(
            item_spacing=kerf,
            strategy="best-fit-decreasing",
        ),
    )

    result_stocks: List[StockResult1D] = []
    total_stock_measure = 0.0
    total_used_measure = 0.0
    total_waste_measure = 0.0
    total_remnant_measure = 0.0
    total_placed_count = 0

    for packed_bin in pack_result.bins:
        total_stock_measure += packed_bin.capacity
        placements: List[Placement1D] = []
        cuts: List[Cut1D] = []

        for i, item in enumerate(packed_bin.items):
            if i > 0:
                cuts.append(Cut1D(
                    x=item.offset - kerf,
                    kerf=kerf,
                    step=i,
                ))
            placements.append(Placement1D(
                part_id=item.id,
                x=item.offset,
                length=item.size,
            ))
            total_used_measure += item.size

        total_placed_count += len(placements)

        remaining = packed_bin.remaining_capacity
        remnants: List[Segment1D] = []
        waste: List[Segment1D] = []

        if remaining > 0:
            if min_remnant_length > 0 and remaining >= min_remnant_length:
                remnants.append(Segment1D(x=packed_bin.used_capacity, length=remaining))
                total_remnant_measure += remaining
            else:
                waste.append(Segment1D(x=packed_bin.used_capacity, length=remaining))
                total_waste_measure += remaining

        cut_loss = len(cuts) * kerf
        total_waste_measure += cut_loss

        result_stocks.append(StockResult1D(
            stock_id=packed_bin.bin_id,
            index=packed_bin.index,
            length=packed_bin.capacity,
            placements=placements,
            cuts=cuts,
            remnants=remnants,
            waste=waste,
        ))

    unplaced_parts = [
        UnplacedPart(part_id=u.id, quantity=u.quantity)
        for u in pack_result.unpacked_items
    ]

    yield_rate = (total_used_measure / total_stock_measure) if total_stock_measure > 0 else 0.0

    return OptimizationResult(
        dimension="1D",
        summary=Summary(
            stock_count_used=len(result_stocks),
            parts_placed=total_placed_count,
            parts_total=pack_result.summary.items_total,
            total_stock_measure=round(total_stock_measure, 4),
            total_used_measure=round(total_used_measure, 4),
            total_waste_measure=round(total_waste_measure, 4),
            total_remnant_measure=round(total_remnant_measure, 4),
            yield_rate=round(yield_rate, 4),
        ),
        stocks=result_stocks,
        unplaced_parts=unplaced_parts,
    )

File: packages/typescript/src/optimizer1d.ts
import {
  InputRequest,
  OptimizationResult,
  Stock1D,
  Part1D,
  StockResult1D,
  Placement1D,
  Cut1D,
  Segment1D,
  UnplacedPart,
} from './types.js';
import { binPack1D, BinDefinition, ItemDefinition } from './binpacking/index.js';

export function optimize1D(input: InputRequest): OptimizationResult {
  const kerf = Math.max(0, input.kerf ?? 0);
  const minRemnantLength = input.min_remnant_size?.length ?? 0;

  const bins: BinDefinition<Stock1D>[] = (input.stocks as Stock1D[]).map((s) => ({
    id: s.id,
    capacity: s.length,
    quantity: s.quantity ?? 1,
    cost: s.cost ?? s.length,
    data: s,
  }));

  const items: ItemDefinition<Part1D>[] = (input.parts as Part1D[]).map((p) => ({
    id: p.id,
    size: p.length,
    quantity: p.quantity ?? 1,
    data: p,
  }));

  const packResult = binPack1D(bins, items, {
    itemSpacing: kerf,
    strategy: 'best-fit-decreasing',
  });

  const resultStocks: StockResult1D[] = [];
  let totalStockMeasure = 0;
  let totalUsedMeasure = 0;
  let totalWasteMeasure = 0;
  let totalRemnantMeasure = 0;
  let totalPlacedCount = 0;

  for (const packedBin of packResult.bins) {
    totalStockMeasure += packedBin.capacity;
    const placements: Placement1D[] = [];
    const cuts: Cut1D[] = [];

    for (let i = 0; i < packedBin.items.length; i++) {
      const item = packedBin.items[i];
      if (i > 0) {
        cuts.push({
          x: item.offset - kerf,
          kerf: kerf,
          step: i,
        });
      }
      placements.push({
        part_id: item.id,
        x: item.offset,
        length: item.size,
      });
      totalUsedMeasure += item.size;
    }

    totalPlacedCount += placements.length;

    const remaining = packedBin.remainingCapacity;
    const remnants: Segment1D[] = [];
    const waste: Segment1D[] = [];

    if (remaining > 0) {
      if (minRemnantLength > 0 && remaining >= minRemnantLength) {
        remnants.push({
          x: packedBin.usedCapacity,
          length: remaining,
        });
        totalRemnantMeasure += remaining;
      } else {
        waste.push({
          x: packedBin.usedCapacity,
          length: remaining,
        });
        totalWasteMeasure += remaining;
      }
    }

    // Cut loss is also considered waste
    const cutLoss = cuts.length * kerf;
    totalWasteMeasure += cutLoss;

    resultStocks.push({
      stock_id: packedBin.binId,
      index: packedBin.index,
      length: packedBin.capacity,
      placements,
      cuts,
      remnants,
      waste,
    });
  }

  const unplaced_parts: UnplacedPart[] = packResult.unpackedItems.map((u) => ({
    part_id: u.id,
    quantity: u.quantity,
  }));

  const yieldRate = totalStockMeasure > 0 ? totalUsedMeasure / totalStockMeasure : 0;

  return {
    dimension: '1D',
    summary: {
      stock_count_used: resultStocks.length,
      parts_placed: totalPlacedCount,
      parts_total: packResult.summary.itemsTotal,
      total_stock_measure: Number(totalStockMeasure.toFixed(4)),
      total_used_measure: Number(totalUsedMeasure.toFixed(4)),
      total_waste_measure: Number(totalWasteMeasure.toFixed(4)),
      total_remnant_measure: Number(totalRemnantMeasure.toFixed(4)),
      yield_rate: Number(yieldRate.toFixed(4)),
    },
    stocks: resultStocks,
    unplaced_parts,
  };
}

